server/bytestream/bytestream.go

```diff
@@ -14,11 +15,140 @@ import (
 	"github.com/buildbuddy-io/buildbuddy/server/util/status"
 	"google.golang.org/protobuf/proto"
 
+	arpb "github.com/buildbuddy-io/buildbuddy/proto/archive"
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
 	bspb "google.golang.org/genproto/googleapis/bytestream"
 )
 
+func FetchBytestreamZipManifest(ctx context.Context, env environment.Env, url *url.URL) (*arpb.ArchiveManifest, error) {
+	r, err := digest.ParseDownloadResourceName(strings.TrimPrefix(url.RequestURI(), "/"))
+	if err != nil {
+		return nil, err
+	}
+	// Let's just read 64K and see if we can find the central directory in there.
+	// We probably don't want to be in the business of rendering 3000-file zips'
+	// contents anyway.
+	offset := r.GetDigest().GetSizeBytes() - 65536
+	if offset < 0 {
+		offset = 0
+	}
+
+	reader, writer := io.Pipe()
+	go func() {
+		err = StreamBytestreamFileChunk(ctx, env, url, offset, r.GetDigest().GetSizeBytes()-offset, func(data []byte) {
+			writer.Write(data)
+		})
+		writer.CloseWithError(err)
+	}()
+
+	// Dump the full contents out into a buffer (should be 64K or less).
+	footer, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find and parse the End of Central Directory header or fail.
+	eocd, err := readDirectoryEnd(footer, r.GetDigest().GetSizeBytes())
+	if err != nil {
+		return nil, err
+	}
+
+	cdStart := eocd.directoryOffset - offset
+	cdEnd := cdStart + eocd.directorySize
+
+	out := &arpb.ArchiveManifest{}
+	entries, err := readDirectoryHeader(footer[cdStart:cdEnd], eocd)
+	out.Entry = entries
+	return out, nil
```

Comment:
you could rewrite this (and check this error) like:

```
entries, err := readDirectoryHeader(footer[cdStart:cdEnd], eocd)
if err != nil {
  return nil, err
}
return &arpb.ArchiveManifest{Entry: entries}, nil
```

Reply:
Gah, sorry, dunno how i missed the check here.